src/obstacle_detection_node.py

```diff
@@ -0,0 +1,264 @@
+
+from constants import *
+
+# ==============================================================================
+# === Global Variables =========================================================
+# ==============================================================================
+Q_MATRIX = None
+DISPARITY_MAP = None
+OBSTACLE_POINTS = None
+DRONE_POSE = None
+
+CAM_INFO1 = None
+CAM_INFO2 = None
+
+MAPX1 = None
+MAPY1 = None
+MAPX2 = None
+MAPY2 = None
+
+OBSTACLE_RADIUS = 0.1524 # meters
+
+BRIDGE = CvBridge()
+
+# ==============================================================================
+# === Helper Functions =========================================================
+# ==============================================================================
+def update_disparity_map(msg):
+    global DISPARITY_MAP
+    DISPARITY_MAP = msg
+
+def update_drone_pose(msg):
+    global DRONE_POSE
+    DRONE_POSE = msg
+
+def camera_info_cb1(msg):
+    global CAM_INFO1
+    CAM_INFO1 = copy.deepcopy(msg)
+
+def camera_info_cb2(msg):
+    global CAM_INFO2
+    CAM_INFO2 = copy.deepcopy(msg)
+
+def init_maps():
+    """
+        Use to obtain the Q matrix for the stereo camera setup
+    """
+    global MAPX1, MAPY1, MAPX2, MAPY2
+
+    K1 = np.array(CAM_INFO1.K).reshape(3,3)
+    D1 = np.array(CAM_INFO1.D)
+    K2 = np.array(CAM_INFO2.K).reshape(3,3)
+    D2 = np.array(CAM_INFO2.D)
+    T = np.array([BASELINE, 0, 0]) # 64 mm baseline
+    R1, R2, P1, P2, Q, roi1, roi2 = cv2.stereoRectify(
+        K1, D1, K2, D2, IMG_SIZE_WH, R=np.eye(3), T=T
+    )
+
+    MAPX1, MAPY1 = cv2.fisheye.initUndistortRectifyMap(
+        K1, D1, R1, P1, size=IMG_SIZE_WH, m1type=cv2.CV_32FC1
+    )
+    MAPX2, MAPY2 = cv2.fisheye.initUndistortRectifyMap(
+        K2, D2, R2, P2, size=IMG_SIZE_WH, m1type=cv2.CV_32FC1
+    )
+
+    global Q_MATRIX
+    Q_MATRIX = Q
+
+
+def process_point_cloud(disparity, Q_MATRIX):
+    """
+    """
+    
+    # Step 1: Reproject disparity map to 3D point cloud
+    point_cloud = cv2.reprojectImageTo3D(disparity, Q_MATRIX)
+
+    # Step 2: Filter out points with depth value less than and greater than thresholds
+    min_threshold = 0.3 # meters
+    max_threshold = 2.5 # meters
+
+    mask = np.logical_and(point_cloud[:,:,2] > min_threshold, point_cloud[:,:,2] < max_threshold)
+    point_cloud = point_cloud[mask]
+
+    # Step 3: Find all points at y = 0
+    mask = point_cloud[:,:,1] == 0
+    point_cloud = point_cloud[mask]
+
+    # Step 4: Find the closest point to the drone
+    min_distance = np.min(point_cloud[:,:,2])
+    mask = point_cloud[:,:,2] == min_distance
+    closest_obstacles = point_cloud[mask]
+
+    # Step 5: Use the coordinate of the closest obstacle to locate obstacle in map frame
+    located_obstacle_pose_stamp = locate_obstacles(closest_obstacles)
+
+    # Step 6: Transform the located obstacle to world frame
+    world_obstacle_pose_stamp = transform_map_to_world(located_obstacle_pose_stamp)
```

Comment:
We can actually do all of the above computation in world space to begin with (since we should have the drone position in world space and we can transform the PC to world space using the TF tree)